app/code/Magento/CatalogUrlRewrite/Plugin/Webapi/Controller/Rest/InputParamsResolver.php

```diff
@@ -41,48 +45,98 @@ public function __construct(RestRequest $request)
      */
     public function afterResolve(\Magento\Webapi\Controller\Rest\InputParamsResolver $subject, array $result): array
     {
-        $route = $subject->getRoute();
-        $serviceMethodName = $route->getServiceMethod();
-        $serviceClassName = $route->getServiceClass();
-        $requestBodyParams = $this->request->getBodyParams();
+        $this->processProductCall($subject, $result);
+        $this->processCategoryCall($subject, $result);
```

Comment:
This class now has more than one reason to change, breaking the SRP. Also looking at the plugin name it references it to product `product_save_rewrites_history_rest_plugin`. Also, the code between product and category seems to be duplication, can we create it in a general way and maybe handle things with DI?